src/utils/logger.rs

```diff
@@ -0,0 +1,197 @@
+        accessible,
+        level: MessageLevel::Info,
+    });
+}
+
+pub fn warning(contents: String) {
+    logger_thread::send_log_message(PrintMessage {
+        contents,
+        // Warnings are important and unlikely to flood, so they should be displayed
+        accessible: true,
+        level: MessageLevel::Warning,
+    });
+}
+
+/// Message object used for sending logs from worker threads to a logging thread via channels.
+/// See <https://github.com/ouch-org/ouch/issues/643>
+#[derive(Debug)]
+struct PrintMessage {
+    contents: String,
+    accessible: bool,
+    level: MessageLevel,
+}
+
+impl PrintMessage {
+    fn to_processed_message(&self) -> Option<String> {
+        match self.level {
+            MessageLevel::Info => {
+                if self.accessible {
+                    if is_running_in_accessible_mode() {
+                        Some(format!("{}Info:{} {}", *YELLOW, *RESET, self.contents))
+                    } else {
+                        Some(format!("{}[INFO]{} {}", *YELLOW, *RESET, self.contents))
+                    }
+                } else if !is_running_in_accessible_mode() {
+                    Some(format!("{}[INFO]{} {}", *YELLOW, *RESET, self.contents))
+                } else {
+                    None
+                }
+            }
+            MessageLevel::Warning => {
+                if is_running_in_accessible_mode() {
+                    Some(format!("{}Warning:{} {}", *ORANGE, *RESET, self.contents))
+                } else {
+                    Some(format!("{}[WARNING]{} {}", *ORANGE, *RESET, self.contents))
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum MessageLevel {
+    Info,
+    Warning,
+}
+
+mod logger_thread {
+    use std::{
+        sync::{mpsc::RecvTimeoutError, Arc, Barrier},
+        time::Duration,
+    };
+
+    use super::*;
+
+    type LogReceiver = mpsc::Receiver<Option<PrintMessage>>;
+    type LogSender = mpsc::Sender<Option<PrintMessage>>;
+
+    static SENDER: OnceLock<LogSender> = OnceLock::new();
+
+    #[track_caller]
+    fn setup_channel() -> LogReceiver {
+        let (tx, rx) = mpsc::channel();
+        SENDER.set(tx).expect("`setup_channel` should only be called once");
+        rx
+    }
+
+    #[track_caller]
+    fn get_sender() -> &'static LogSender {
+        SENDER.get().expect("No sender, you need to call `setup_channel` first")
+    }
+
+    #[track_caller]
+    pub(super) fn send_log_message(msg: PrintMessage) {
+        send_message(Some(msg));
+    }
+
+    #[track_caller]
+    fn send_message(msg: Option<PrintMessage>) {
+        get_sender().send(msg).expect("Failed to send internal message");
+    }
+
+    pub struct LoggerThreadHandle {
+        shutdown_barrier: Arc<Barrier>,
+    }
+
+    impl LoggerThreadHandle {
+        /// Tell logger to shutdown and waits till it does.
+        pub fn shutdown_and_wait(self) {
+            // Signal the shutdown
+            send_message(None);
```

Comment:
> Ah yes just one thing I wanted to comment on was that maybe instead of sending a None message to indicate a flush we could have an enum for clarity but like you are the maintainer so if it makes sense for you keep it 😅

Yes!!!!! Thank you.

I deleted my notes by accident and forgot something but couldn't remember what it was. There it is. Yeeeess I'll do it for sure.

Reply:
I just changed this, let me know if the new version is fine :)